text/0000-time-series-gem-downloads.md

```diff
@@ -0,0 +1,205 @@
+and find_or_create a Download record for that version (with its version ID and rubygem ID) with the download count, log ticket ID, and time bucket.
+
+Each continuous aggregate will get its own subclass of `Download`, so we are able to query agains the matrix of timeframes x [single gem version, all versions for a gem, all gems] using a consistent interface.
+
+For example, all of the following will be possible (and queriable scoped to particular timeframes, gems, gem versions, etc.)
+
+```ruby
+irb(main):004:0> Download::P1MAllVersion.all
+2023-04-03 10:11:48.902642 D [920:8000 log_subscriber.rb:130] ActiveRecord::Base --   Download::P1MAllVersion Load (173.3ms)  SELECT "downloads_1_month_all_versions".* FROM "downloads_1_month_all_versions"
+=>
+[#<Download::P1MAllVersion:0x000000011cfd37a8 occurred_at: Sun, 01 Nov 2015 00:00:00.000000000 UTC +00:00, rubygem_id: 11, downloads: 6>,
+ #<Download::P1MAllVersion:0x000000011cfd3668 occurred_at: Tue, 01 Dec 2015 00:00:00.000000000 UTC +00:00, rubygem_id: 11, downloads: 1>]
+irb(main):005:0> Download::P1YAllGem.all
+2023-04-03 10:11:57.916116 D [920:8000 log_subscriber.rb:130] ActiveRecord::Base --   Download::P1YAllGem Load (75.0ms)  SELECT "downloads_1_year_all_gems".* FROM "downloads_1_year_all_gems"
+=> [#<Download::P1YAllGem:0x000000011c6de6c0 occurred_at: Thu, 01 Jan 2015 00:00:00.000000000 UTC +00:00, downloads: 7>]
+irb(main):006:0> Download::P1D.all
+2023-04-03 10:12:01.759285 D [920:8000 log_subscriber.rb:130] ActiveRecord::Base --   Download::P1D Load (43.5ms)  SELECT "downloads_1_day".* FROM "downloads_1_day"
+=>
+[#<Download::P1D:0x000000011cef9b20 occurred_at: Mon, 30 Nov 2015 00:00:00.000000000 UTC +00:00, rubygem_id: 11, version_id: 21, downloads: 2>,
+ #<Download::P1D:0x000000011cef9a80 occurred_at: Wed, 30 Dec 2015 00:00:00.000000000 UTC +00:00, rubygem_id: 11, version_id: 23, downloads: 1>,
+ #<Download::P1D:0x000000011cef99e0 occurred_at: Mon, 30 Nov 2015 00:00:00.000000000 UTC +00:00, rubygem_id: 11, version_id: 22, downloads: 1>,
+ #<Download::P1D:0x000000011cef9940 occurred_at: Sun, 29 Nov 2015 00:00:00.000000000 UTC +00:00, rubygem_id: 11, version_id: 23, downloads: 1>,
+ #<Download::P1D:0x000000011cef98a0 occurred_at: Mon, 30 Nov 2015 00:00:00.000000000 UTC +00:00, rubygem_id: 11, version_id: 23, downloads: 2>]
+```
+
+Note that the raw `Download` class will not be exposed anywhere -- as far as the rails app is concerned, it's a write-only table, and all queries will happen against the continuous aggregates.
+
+# Drawbacks
+
+- Adding another datastore complicates contributing to the application
+- Another datastore means we can't natively write joins against the download data
+- Timescale cloud costs money
+- There is no terraform provider for timescale cloud (yet) requiring us to write one or manually provision
+- This is still high cardinality data, and there's no knowing for sure how well timescale will compress it
```

Comment:
I can help testing and benchmarking it. Is there any data available to seed the database? Or if anyone can share any numbers to give me some insights about how big is the cardinality.